ts/test/local/scenario.ts

```diff
@@ -0,0 +1,146 @@
+import {
+  AppSignal,
+  AppSignalCb,
+  DnaSource,
+  EntryHash,
+} from "@holochain/client";
+import test from "tape-promise/tape";
+import { LocalScenario } from "../../src/local/scenario";
+import { pause } from "../../src/util";
+import { FIXTURE_DNA_URL, FIXTURE_HAPP_URL } from "../fixture";
+
+test("Local Scenario - Install hApp bundle and access cells through role ids", async (t) => {
+  const scenario = new LocalScenario();
+  const alice = await scenario.addPlayerWithHappBundle({
+    path: FIXTURE_HAPP_URL.pathname,
+  });
+  t.ok(alice.namedCells.get("test"));
+
+  await scenario.cleanUp();
+});
```

Comment:
The implications are indeed that the test will hang, because the conductors are still running. You don't have to call `cleanUp`, but `shutDown`.

I find the syntax of closures more difficult to read than "sequential" (what is the right word here?) code, but I see the benefit of your suggestion. You suggest to provide this in addition to instantiating a Scenario, right?

Reply:
In the "sequential" approach, what happens if an exception is thrown before `cleanUp`? Will the test hang or fail properly? It seems like the sequential approach needs to be instead
```ts
const scenario = new LocalScenario();
try {
  /* use scenario */
} finally (
  await secnario.cleanUp()
}
```

Reply:
Yes, the benefit of the closure approach is that all boilerplate can be moved into the function which calls the closure, and the closure can be limited just to the part the test writer actually cares to test. 

I'd suggest writing tests with exceptions being intentionally from various places to ensure they're handled properly. A lot of work was put into that for the previous version. I feel it is our fault if a test writer writes a test that unintentionally hangs, not theirs.

Reply:
It happens with several test runners that a test hangs when an unhandled exception occurs. Other test runners automatically finish if nothing happens after a given interval.

Further it's common that tests hang or fail while being developed and have to be debugged. This is not exclusive to our framework.

Yet I agree and will write some tests to ensure that the wrapper handles those exceptions. During development I've checked several error sources, also for TryCP which requires custom error handling.

Reply:
In Holo's code, we usually use Jest as our test runner, which supports `beforeAll` and `afterAll` declarations for test spinup and teardown. We would do something like
```js
let scenario

beforeAll(async () => {
  scenario = new LocalScenario()
  // Do other setup code using `scenario`
})
afterAll(async () => {
  await scenario.cleanUp()
})
```

Writing it with closures (which we end up doing in the current tryorama API), would result in

```js
let scenario
let endScenario
let scenarioCleanedUp

beforeAll(async () => {
    let scenarioStarted
    const scenarioPromise = new Promise(resolve => (scenarioStarted = resolve))
    const scenarioEndedPromise = new Promise(resolve => (endScenario = resolve))
    scenarioCleanedUp = localScenario(async s => {
      scenarioStarted(s)
      await scenarioEndedPromise
    })
    scenario = await scenarioPromise
  // Do other setup code using `scenario`
})

afterAll(() => {
  endScenario()
  await scenarioCleanedUp
})
```

So let's at least keep the imperative version :)

(See [here](https://github.com/Holo-Host/holochain-scalability-tests/blob/5f1758ce19ff3180cb1dcd9086eb2a3e83ac40b9/tests/holofuel/scaletest.js#L45-L53) for a real example of this)

Reply:
(Responding to https://github.com/holochain/tryorama/pull/108#discussion_r871555181)

Remove the
```
  } catch (error) {
    console.error(error);
  }
```
And then that looks right.

Note that `try { ... } finally { ... }` works without a catch. Also note that we should really not be using `console.error` in tryorama, because in a NodeJS context it's the same as `console.log` except that it prints to stderr instead of stdout. And in a Jest context, it actually causes crashes. ([We had to remove a `console.error` from holochain-client-js for this reason](https://github.com/holochain/holochain-client-js/pull/85))

Reply:
> (Responding to [#108 (comment)](https://github.com/holochain/tryorama/pull/108#discussion_r871555181))
> 
> Remove the
> 
> ```js
>   } catch (error) {
>     console.error(error);
>   }
> ```
> 
> And then that looks right.
> 
> Note that `try { ... } finally { ... }` works without a catch. Also note that we should really not be using `console.error` in tryorama, because in a NodeJS context it's the same as `console.log` except that it prints to stderr instead of stdout. And in a Jest context, it actually causes crashes. ([We had to remove a `console.error` from holochain-client-js for this reason](https://github.com/holochain/holochain-client-js/pull/85))

I added the `catch` block because otherwise the developer would have to add another `try..catch` around the `runScenario` call to actually catch the error. I have used `logger.error` now, and ran a failing test with Jest.

It works, but I need to output `logger.error(JSON.stringify(error))`, otherwise I just get `[object Object]`.

Reply:
> I'd suggest writing tests with exceptions being intentionally thrown from various places to ensure they're handled properly. A lot of work was put into that for the previous version. I feel it is our fault if a test writer writes a test that unintentionally hangs, not theirs.

@maackle I've added a few tests to check if errors are caught and logged and the tests finish. I couldn't think of more scenarios that could go wrong. The local conductor is fairly straightforward, for TryCP there'd be many more cases to test.